tests: e2e test `graceshutdown` failed

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

Run CI pipeline [merged_e2e_test](https://prow.tidb.net/?repo=pingcap%2Ftidb&type=postsubmit&job=pingcap%2Ftidb%2Fmerged_e2e_test).

### 2. What did you expect to see? (Required)

Pipeline succeed.

### 3. What did you see instead (Required)

Pipeline was failed.

```

+ cd tests/graceshutdown
+ make
cd /home/jenkins/agent/workspace/pingcap/tidb/merged_e2e_test/tidb && \
CGO_ENABLED=1 GO111MODULE=on go build  -tags codes -ldflags '-X "github.com/pingcap/tidb/parser/mysql.TiDBReleaseVersion=v7.3.0-alpha-55-g767267a184" -X "github.com/pingcap/tidb/util/versioninfo.TiDBBuildTS=2023-06-27 02:37:24" -X "github.com/pingcap/tidb/util/versioninfo.TiDBGitHash=767267a1845a4d9bfab7835142bc9beb9f491489" -X "github.com/pingcap/tidb/util/versioninfo.TiDBGitBranch=HEAD" -X "github.com/pingcap/tidb/util/versioninfo.TiDBEdition=Community"' -o /home/jenkins/agent/workspace/pingcap/tidb/merged_e2e_test/tidb/tests/graceshutdown/bin/tidb-server tidb-server/main.go
Build TiDB Server successfully!
+ ./run-tests.sh
go: downloading go.uber.org/goleak v1.2.1
[2023/06/27 10:37:44.307 +08:00] [INFO] [graceshutdown_test.go:48] ["starting tidb"] [cmd="bin/tidb-server --store=mocktikv --path=/tmp/tidb_gracefulshutdown/mocktikv -P=5501 --status=8500 --log-file=/tmp/tidb_gracefulshutdown/tidb5501.log"]
[2023/06/27 10:37:44.808 +08:00] [WARN] [graceshutdown_test.go:89] ["ping addr failed"] [addr=127.0.0.1:5501] ["retry count"=0] [error="dial tcp 127.0.0.1:5501: connect: connection refused"]
[2023/06/27 10:37:45.059 +08:00] [WARN] [graceshutdown_test.go:89] ["ping addr failed"] [addr=127.0.0.1:5501] ["retry count"=1] [error="dial tcp 127.0.0.1:5501: connect: connection refused"]
[2023/06/27 10:37:45.561 +08:00] [INFO] [graceshutdown_test.go:113] ["connect to server ok"] [addr=127.0.0.1:5501]
[2023/06/27 10:37:46.583 +08:00] [INFO] [graceshutdown_test.go:61] ["service Interrupt"] [name=tidb]
[mysql] 2023/06/27 10:37:46 packets.go:37: unexpected EOF
--- FAIL: TestGracefulShutdown (2.28s)
    graceshutdown_test.go:155: 
        	Error Trace:	/home/jenkins/agent/workspace/pingcap/tidb/merged_e2e_test/tidb/tests/graceshutdown/graceshutdown_test.go:155
        	Error:      	Received unexpected error:
        	            	invalid connection
        	Test:       	TestGracefulShutdown
FAIL
exit status 1
FAIL	github.com/pingcap/tidb/tests/graceshutdown	2.280s
script returned exit code 1
```

See https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fmerged_e2e_test/detail/merged_e2e_test/1588/pipeline/.

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/tests/graceshutdown/graceshutdown_test.go
+++ b/tests/graceshutdown/graceshutdown_test.go
@@ -131,7 +131,9 @@ func TestGracefulShutdown(t *testing.T) {
 	conn1, err := db.Conn(ctx)
 	require.NoError(t, err)
 	defer func() {
-		require.NoError(t, conn1.Close())
+		if conn1 != nil {
+			require.NoError(t, conn1.Close())
+		}
 	}()
 
 	_, err = conn1.ExecContext(ctx, "drop table if exists t;")
@@ -149,10 +151,19 @@ func TestGracefulShutdown(t *testing.T) {
 		close(done)
 	}()
 
+	// Graceful shutdown will wait for connections in transaction only.
+	// See https://github.com/pingcap/tidb/pull/44953.
+	txn, err := conn1.BeginTx(ctx, nil)
+	require.NoError(t, err)
 	sql := `select 1 from t where not (select sleep(3)) ;`
 	var a int64
-	err = conn1.QueryRowContext(ctx, sql).Scan(&a)
+	err = txn.QueryRowContext(ctx, sql).Scan(&a)
 	require.NoError(t, err)
 	require.Equal(t, a, int64(1))
+	require.NoError(t, txn.Commit())
+
+	conn1.Close()
+	conn1 = nil
+
 	<-done
 }

